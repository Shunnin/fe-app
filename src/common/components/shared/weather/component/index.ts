export { Temperature } from './temperature/temperature.component';
export { WeatherCard } from './weather-card/weather-card.component';
export { WeatherItem } from './weather-item/weather-item.component';
export * from './weather-icon';
