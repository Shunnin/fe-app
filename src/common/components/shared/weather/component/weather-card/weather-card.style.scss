@import 'style/foundation/config';

.weather-card {
	background-color: $c-wh;
  box-shadow: 1px 1px 3px rgba(0, 0, 0, 0.1);
  border-radius: $b-border-radius-md;
  padding: $b-space 0;
  display: flex;
  flex-direction: column;

	@include respondto(xs, sm) {
		padding: $b-space;
	}

	&__title {
		font-weight: $b-font-weight-medium;
    font-size: $b-line-height-heading;
		margin: $b-space $b-space-2x;

		@include respondto(xs, sm) {
			margin: $b-space 0;
		}
	}

	&__toggle {
		line-height: $b-line-height-heading + 2 * $b-space;
		margin: 0 $b-space-2x;

		@include respondto(xs, sm) {
			margin: 0;
		}
	}

	&__content {
		display: flex;
		flex-wrap: wrap;
		justify-content: center;
	}

	&__status {
		display: flex;
		flex-direction: column;
		margin: $b-space $b-space-2x;
		width: 400px;

	  @include respondto(xs, sm) {
			margin: $b-space 0;
		}

		&-name {
			font-weight: $b-font-weight-bold;
			font-size: $b-font-size-xl;
			color: $c-lb7;
			margin-bottom: $b-space;
		}

		&-desc {
			font-size: $b-font-size-md;
			text-align: center;
			color: $c-gr1;
		}

		.temperature {
			font-weight: $b-font-weight-light;
			font-size: 7 * $b-font-size;
			color: $c-lb7;
			margin-left: $b-space-2x;

			sup {
				line-height: 0;
			}
		}
	}

	&__info {
		display: flex;
		flex-direction: column;
		margin: $b-space $b-space-2x;

		@include respondto(xs, sm) {
			margin: $b-space 0;
		}

		&-feels-like {
			font-size: $b-font-size-md;
  		color: $c-gr1;
		}

		&-temp-degree{
			display: flex;
			margin-bottom: $b-space;
			width: 16px;

			.weather-degree {
				display: flex;
				align-items: center;
				font-weight: $b-font-weight-medium;
				font-size: $b-font-size-md;
				color: $c-lb7;

				margin-top: $b-space;
				margin-right: $b-space-3x;

				svg {
					fill: $c-lb1;
					margin-right: $b-space-s;
				}
			}
		}

		&-detail {
			display: flex;
			align-items: center;
			justify-content: flex-start;
			margin-bottom: $b-space;

			div {
				color: $c-gr1;
				display: flex;
				align-items: center;
				font-size: $b-font-size;
				width: 128px;
			}

			svg {
				fill: $c-lb1;
				width: 26px;
				margin-right: $b-space-s;
			}

			span {
				color: $c-lb7;
				font-weight: $b-font-weight-medium;
				font-size: $b-font-size;
			}
		}
	}
}
