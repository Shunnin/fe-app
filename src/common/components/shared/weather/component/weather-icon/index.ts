export { WeatherIcon } from './weather-icon.component';
export * from './weather-icon.constant';
