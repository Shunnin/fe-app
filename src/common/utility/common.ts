export const getNextSevenDays = (): string[] => {
  const days = ['Sun', 'Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat'];
  const next7Days = [];

  // eslint-disable-next-line no-magic-numbers
  for (let i = 0; i < 7; i++) {
    // eslint-disable-next-line no-magic-numbers
    next7Days.push(days[new Date(Date.now() + (i + 1) * 24 * 60 * 60 * 1000).getDay()]);
  }

  return next7Days;
};
