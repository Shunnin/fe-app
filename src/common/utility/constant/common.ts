import { noop, transform } from 'lodash-es';

export const TEMP_UNIT = Object.freeze({
  CELSIUS: 'CELSIUS',
  FAHRENHEIT: 'FAHRENHEIT',
});

export const TEMP_UNIT_VALUE = Object.freeze({
  [TEMP_UNIT.CELSIUS]: true,
  [TEMP_UNIT.FAHRENHEIT]: false,
});

export const TEMP_UNIT_TOGGLE_VALUE = transform(TEMP_UNIT_VALUE, (result, value, key) => (result[value] = key));

export const TEMP_UNIT_MAPPING_LABEL = Object.freeze({
  true: 'C',
  false: 'F',
});

export const EMPTY_FUNC = noop;
export type IFunction = ReturnType<typeof EMPTY_FUNC>;

export const DEFAULT_LOCALE = 'en';
