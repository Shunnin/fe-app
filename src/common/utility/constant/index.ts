export * from './number';
export * from './common';
export * from './url';
