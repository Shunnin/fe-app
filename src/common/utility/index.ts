export * from './route';
export * from './constant';
export * from './promise';
export * from './temperature';
export * from './common';
export * from './url';
