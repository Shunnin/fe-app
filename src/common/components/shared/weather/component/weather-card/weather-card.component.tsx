import { FC, memo } from 'react';
import { FormattedMessage, useIntl } from 'react-intl';
import { isEmpty } from 'lodash-es';

import { WeatherData } from '../../../../../../models';
import { EMPTY_FUNC, TEMP_UNIT, TEMP_UNIT_MAPPING_LABEL, TEMP_UNIT_VALUE, kmToMile } from '../../../../../utility';
import { ToggleSwitch } from '../../../../core/toggle-switch';

import { Temperature } from '../temperature/temperature.component';
import { WeatherIcon, HighIcon, HumidityIcon, LowIcon, PressureIcon, WindIcon } from '../weather-icon';

import './weather-card.style';

interface IWeatherCardStatusProps {
  weatherCode: number;
  name: string;
  temp: number;
  description: string;
}

const WeatherCardStatus: FC<IWeatherCardStatusProps> = ({ weatherCode, name, temp, description }) => {
  return (
    <div className="weather-card__status">
      <div className="weather-card__status-name">{name}</div>
      <div className="flex">
        <WeatherIcon code={weatherCode} big />
        <Temperature value={temp} />
      </div>
      <div className="weather-card__status-desc text-capitalize">{description}</div>
    </div>
  );
};

interface IWeatherCardInfoProps {
  weather: WeatherData | {};
  degreeType: string;
}

const WeatherCardInfo: FC<IWeatherCardInfoProps> = ({ weather, degreeType }) => {
  const { tempDetails: { feelsLike, tempMax, tempMin, humidity, pressure } = {}, wind: { speed } = {} } =
    weather as WeatherData;

  return (
    <div className="weather-card__info">
      <div className="weather-card__info-feels-like">
        <FormattedMessage
          defaultMessage="Feels like {feelsLike}"
          values={{ feelsLike: <Temperature value={feelsLike} /> }}
        />
      </div>
      <div className="weather-card__info-temp-degree">
        <div className="weather-degree">
          <HighIcon />
          <Temperature value={tempMax} />
        </div>
        <div className="weather-degree">
          <LowIcon />
          <Temperature value={tempMin} />
        </div>
      </div>
      <div className="weather-card__info-detail">
        <div>
          <HumidityIcon />
          <FormattedMessage defaultMessage="Humidity" />
        </div>
        <span>
          <FormattedMessage defaultMessage="{humidity}%" values={{ humidity }} />
        </span>
      </div>
      <div className="weather-card__info-detail">
        <div>
          <WindIcon />
          <FormattedMessage defaultMessage="Wind" />
        </div>
        <span>
          {degreeType === TEMP_UNIT.CELSIUS ? (
            <FormattedMessage defaultMessage="{speed} kph" values={{ speed }} />
          ) : (
            <FormattedMessage defaultMessage="{speed} mph" values={{ speed: kmToMile(speed) }} />
          )}
        </span>
      </div>
      <div className="weather-card__info-detail">
        <div>
          <PressureIcon />
          <FormattedMessage defaultMessage="Pressure" />
        </div>
        <span>
          <FormattedMessage defaultMessage="{pressure} hPa" values={{ pressure }} />
        </span>
      </div>
    </div>
  );
};

interface IWeatherCardProps {
  weather: WeatherData | {};
  degreeType: string;
  onClickChangeTempUnit: Function;
}

export const WeatherCard: FC<IWeatherCardProps> = memo(
  ({ weather, degreeType, onClickChangeTempUnit = EMPTY_FUNC }) => {
    const intl = useIntl();
    const { name, id, description, tempDetails: { temp } = {} } = weather as WeatherData;

    if (isEmpty(weather)) {
      return <></>;
    }

    return (
      <div className="weather-card">
        <div className="flex space-between">
          <div className="weather-card__title">
            <FormattedMessage defaultMessage="Current Weather" />
          </div>
          <div className="weather-card__toggle">
            <ToggleSwitch
              name={intl.formatMessage({ defaultMessage: 'Temperature Unit' })}
              defaultValue={TEMP_UNIT_VALUE.CELSIUS}
              mappingLabel={TEMP_UNIT_MAPPING_LABEL}
              onClick={onClickChangeTempUnit}
            />
          </div>
        </div>
        <div className="weather-card__content">
          <WeatherCardStatus weatherCode={id} name={name} temp={temp} description={description} />
          <WeatherCardInfo weather={weather} degreeType={degreeType} />
        </div>
      </div>
    );
  },
);
WeatherCard.displayName = 'WeatherCard';
