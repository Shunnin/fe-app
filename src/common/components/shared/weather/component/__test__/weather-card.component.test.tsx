import { render } from '@testing-library/react';

import { TEMP_UNIT } from '../../../../../utility/constant';
import { createMockStore, withIntlProvider } from '../../../../../utility/test';

import { WeatherCard } from '../weather-card/weather-card.component';

describe('<WeatherCard />', () => {
  let store = null;
  const onClickChangeTempUnitMock = jest.fn();
  const validWeather = {
    description: 'scattered clouds',
    forecast: 'Clouds',
    icon: '03d',
    id: 802,
    name: 'Ho Chi Minh City',
    tempDetails: {
      feelsLike: 36,
      humidity: 49,
      pressure: 1006,
      temp: 33,
      tempMax: 33,
      tempMin: 33,
    },
    wind: {
      deg: 170,
      speed: 11,
    },
  };
  const props = {
    weather: validWeather,
    degreeType: TEMP_UNIT.CELSIUS,
    onClickChangeTempUnit: onClickChangeTempUnitMock,
  };

  beforeAll(() => {
    store = createMockStore();
  });

  it('renders empty element and matches the snapshot when given empty weather', () => {
    const { container } = render(withIntlProvider(store)(<WeatherCard {...props} weather={{}} />));

    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders weather card and matches the snapshot when given valid weather', () => {
    const { container } = render(withIntlProvider(store)(<WeatherCard {...props} />));

    expect(container.firstChild).toMatchSnapshot();
  });
});
