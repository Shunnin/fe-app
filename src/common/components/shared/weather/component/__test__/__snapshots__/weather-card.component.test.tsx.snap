// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<WeatherCard /> renders empty element and matches the snapshot when given empty weather 1`] = `null`;

exports[`<WeatherCard /> renders weather card and matches the snapshot when given valid weather 1`] = `
<div
  class="weather-card"
>
  <div
    class="flex space-between"
  >
    <div
      class="weather-card__title"
    >
      Current Weather
    </div>
    <div
      class="weather-card__toggle"
    >
      <div
        aria-label="Temperature Unit"
        class="toggle-switch clickable"
        role="button"
        tabindex="0"
      >
        <span
          class="toggle-switch__on"
        >
          C
        </span>
        <div
          class="toggle-switch__indicator is-toggled"
        />
      </div>
    </div>
  </div>
  <div
    class="weather-card__content"
  >
    <div
      class="weather-card__status"
    >
      <div
        class="weather-card__status-name"
      >
        Ho Chi Minh City
      </div>
      <div
        class="flex"
      >
        <reactcomponent
          style="width: 100px; height: 100px;"
        />
        <span
          class="temperature"
        >
          33
          <sup>
            째
          </sup>
        </span>
      </div>
      <div
        class="weather-card__status-desc text-capitalize"
      >
        scattered clouds
      </div>
    </div>
    <div
      class="weather-card__info"
    >
      <div
        class="weather-card__info-feels-like"
      >
        Feels like 
        <span
          class="temperature"
        >
          36
          <sup>
            째
          </sup>
        </span>
      </div>
      <div
        class="weather-card__info-temp-degree"
      >
        <div
          class="weather-degree"
        >
          <reactcomponent />
          <span
            class="temperature"
          >
            33
            <sup>
              째
            </sup>
          </span>
        </div>
        <div
          class="weather-degree"
        >
          <reactcomponent />
          <span
            class="temperature"
          >
            33
            <sup>
              째
            </sup>
          </span>
        </div>
      </div>
      <div
        class="weather-card__info-detail"
      >
        <div>
          <reactcomponent />
          Humidity
        </div>
        <span>
          49%
        </span>
      </div>
      <div
        class="weather-card__info-detail"
      >
        <div>
          <reactcomponent />
          Wind
        </div>
        <span>
          11 kph
        </span>
      </div>
      <div
        class="weather-card__info-detail"
      >
        <div>
          <reactcomponent />
          Pressure
        </div>
        <span>
          1006 hPa
        </span>
      </div>
    </div>
  </div>
</div>
`;
